\documentclass[twocolumn,10pt]{article}

\usepackage[a4paper,hmargin=1.5cm,vmargin=2cm]{geometry}
\setlength{\columnsep}{0.7cm}
%\usepackage{microtype} % nice typesetting  % doesn't work for me :-( -S
\usepackage{amsmath}
\usepackage{graphicx}

\usepackage[utf8]{inputenc}
\usepackage{hyperref}
\urlstyle{same} % no tt font for URLs
\usepackage{natbib}
\bibliographystyle{genome_research}
\setlength{\bibsep}{0.3ex plus 0.2ex}
%\setcitestyle{aysep={}} 
\usepackage[dvipsnames]{xcolor}
\hypersetup{
    colorlinks,
    linkcolor={blue!50!black},
    citecolor={blue!50!black},
    urlcolor={blue!50!black}
}
\renewcommand{\topfraction}{0.9}
\renewcommand{\dbltopfraction}{0.9}
\renewcommand{\textfraction}{0.1}
\clubpenalty=1000
\widowpenalty=1000
\displaywidowpenalty=1000

\usepackage[compact]{titlesec}

\newcommand\blfootnote[1]{%
    \begingroup
    \renewcommand\thefootnote{}\footnote{#1}%
    \addtocounter{footnote}{-1}%
    \endgroup
}

\newcommand{\todo}[1]{[\textcolor{orange}{#1}]}

\begin{document}

\setcounter{secnumdepth}{0}

\twocolumn[{%
\centering
\textbf{\Large Analyzing single-cell bisulfite sequencing data with \texttt{scbs}}\\[1.5ex]

Lukas P.~M. Kremer\textsuperscript{1,2,*}, Martina Braun\textsuperscript{2}, Leonie Küchenhoff\textsuperscript{1,2}, Santiago Cerrizuela\textsuperscript{2},\\Ana Martin-Villalba\textsuperscript{2}, Simon Anders\textsuperscript{1,*}\\[1ex]
{\footnotesize \textsuperscript{1} BioQuant Centre, University of Heidelberg, Heidelberg, Germany\\[-.5ex]
\textsuperscript{2} Molecular Neurobiology Dept., German Cancer Research Center (DKFZ), Heidelberg, Germany}
\vspace{1.5ex}

January 2023
\vspace{5.5ex}

\begin{minipage}{0.7\textwidth}
\small
\textit{Abstract} -- Single-cell bisulfite sequencing (scBS) allows for assessing DNA methylation at single-base pair and single-cell resolution.
Analyzing the big data sets thus obtained requires preprocessing to reduce data size, improve signal over noise, and provide interpretability.
Typically, this is simply done by dividing the genome into large tiles and averaging in these.
We show that this coarse-graining dilutes signal, and suggest improved approaches to determine more informative regions to quantify methylation in, and a better quantitation method than simple averaging.
We demonstrate how our approach enables better distinguishing of cell types and reduces demands on cell number.
We also implement an approach to detect differentially methylated regions (DMRs) between groups of cells.
Finally, we present a software tool, \texttt{scbs}, that implements these methods and offers further functionality required to work with scBS data.
\end{minipage}
\vspace{6.5ex}
}]

\section{Introduction}

Sequencing-based assays with single-cell resolution have offered new means to understand the differences between the cells making up a sample.
Single-cell RNA sequencing (scRNA-seq) techniques have matured at great pace in recent years, with well developed analysis methodology, and methods to study epigenetics at single-cell resolution are rapidly catching up.

\blfootnote{\hspace{-.5cm}\raggedright\textsuperscript{*} e-mail:  \texttt{l.kremer@dkfz-heidelberg.de} and\\ \hspace{1cm}\texttt{simon.anders@bioquant.uni-heidelberg.de}}

Here, we discuss bisulfite sequencing, a method to study DNA methylation.
In brief, DNA is treated with bisulfite, which converts unmethylated cytosines to uracils which are read as thymine in subsequent PCR, while methylated cytosines are protected from conversion.
After sequencing, these conversions allow to determine the methylation status of all cytosines covered by reads \citep{Frommer_1992}.
Bisulfite sequencing can also be performed at single-cell resolution \citep{Smallwood_2014} and even in parallel with scRNA-seq \citep{scMTseq,Clark2018}.

In the present paper, we discuss strategies to analyze single-cell bisulfite-sequencing (scBS) data, suggest improvements to current approaches, show their value in a benchmark, and present software to perform the improved analysis methods.

\phantomsection  % idk why we need this to silence hyperref's warning
\subsection{Standard approach}

\begin{figure*}[t]
    \begin{center}
        \includegraphics[width=0.7\textwidth]{figures/Fig_residuals_AB.png}\\
    \end{center}
    \caption{\small \textbf{Improved quantification of DNA methylation in genomic intervals:}
    \textbf{(A)} Depicted is a genomic interval along a chromosome, for which DNA methylation is to be quantified.
Two cells cover differing parts of the interval with one read each.
If one simply counts for each cell which fraction of the covered CpG sites are methylated, one obtains very different values for the two cells.
    \textbf{(B)} By averaging each CpG site's methylation over all cells and subsequent smoothing, the thick black ``average methylation curve'' is obtained.
To quantify the methylation of cell 1 from (A) relative to this average over all cells, we propose to use the cell's residuals to the smoothed curve (lengths of the vertical red lines) and take their average, counting residuals of methylated CpGs positive and residuals of unmethylated CpGs negative.}
    \label{fig:smoothres}
\end{figure*}


The standard approach to analyze scBS-Seq data is based on methodology developed for the analysis of single-cell RNA-Seq data.
Therefore, we start by briefly reviewing how single-cell \emph{RNA}-Seq data is commonly analyzed, before we discuss scBS data analysis.

The starting point in most scRNA-Seq analyses is a matrix of UMI counts (i.e.
of counts of distinct RNA molecules), one row for each cell and one column for each gene.
A first goal is usually to assign cell types or states to cells.
To this end, one needs to establish which cells are similar to each other, i.e., quantify the distance (i.e., dissimilarity) between any two given cells' transcriptional profile.
A standard approach, used with minor variation in virtually all recent research and automated by popular software such as Seurat \citep{Hao_2021} or ScanPy \citep{Wolf_2018}, is as follows: One first accounts for cell-to-cell variation in sequencing depth by dividing each UMI count by the respective cell's total UMI count, then transforms to a homoskedastic scale by taking the logarithm.
In order to avoid matrix elements with zero count to be transformed to minus infinity, one typically adds a very small ``pseudocount'' (often $10^{-4}$) to the normalized fractions before taking the logarithm.
Now, one could use Euclidean distances of these vectors of logarithmized fractions as dissimilarity score.
However, these scores would be exceedingly noisy due to the strong Poisson noise introduced by the many genes with very low counts.
Therefore, one performs a principal component analysis (PCA), keeping only the top few (typically, 20 to 50) components.
As Poisson noise is uncorrelated between genes, it will average out in the top principal components, as these are all linear combinations with weight on a large number of genes.
Therefore, Euclidean distances between these ``PCA space'' vectors provide a robust dissimilarity score.
Hence, the PCA space representation is suitable as input to methods like t-SNE and UMAP, which provide a two-dimensional representation of the data, or to methods for clustering (assigning cells to groups by similarity) and trajectory finding (identifying elongated manifolds in PCA space and assigning cells to quasi-1D positions along them).

This procedure is commonly adapted when working with single-cell DNA methylation data, so that established methods used in scRNA-Seq can be used.
The task is therefore how to construct a suitable matrix as input for PCA.
A simple and common approach, used for instance by \citet{luo2017single}, is to divide the genome into tiles of e.g. 100~kb size, and calculate for each cell the average methylation of the DNA within each tile.
To this end, we identify all CpG sites in the tile for which we have coverage with at least one read and can hence call the CpG to be either unmethylated or methylated in the cell.
Then we denote as average DNA methylation of this tile in a given cell the proportion of observed CpG sites in the tile that were found to be methylated (Figure \ref{fig:smoothres}A).
This yields a matrix, with one row for each cell and one column for each genomic tile, comprising numbers (methylation fractions) between 0 and 1.
This matrix is now subjected to PCA.
After PCA, one can proceed with dimensionality reduction and clustering approaches known from scRNA-seq.

While this simple procedure is straight-forward and produces satisfactory results, the coarse resolution of large genomic tiles and the sparse read coverage of single-cell methylomes still pose significant challenges in the analysis of scBS data.
Here, we propose three improvements that, as we show afterwards, considerably increase the information content of the processed data.
We will first explain the proposed improvement in a qualitative manner, while mathematical details are given later in the Methods section.
Then, we will demonstrate the improvement using benchmarks.

\section{Improved processing of scBS data}

\phantomsection
\subsection{Read-position aware quantitation} \label{residuals}

We first discuss the task of quantifying the level of methylation in a given, fixed, genomic interval.
In Figure \ref{fig:smoothres}A, such an interval is covered by a single read for two of the three cells shown.
The read from cell 2 shows much more methylation than the read from cell 1, and the standard analysis would therefore consider that cell as having higher methylation in the interval.
However, given that the two reads agree wherever they overlap, a more parsimonious interpretation would be that the cells do not show difference in methylation within the interval.
Rather, both cells, and similarly maybe most other cells, might have stronger methylation in the left third of the interval than in the middle one.

Therefore, we propose to first obtain, for each CpG position, an average of the methylation across all cells, and then quantify for each individual cell its deviation from this average.
In Figure \ref{fig:smoothres}B, the curved line depicts such an average over all cells, and the red vertical lines show an individual cell's deviation from the ensemble average.
We take the lengths of the red lines as signed values (``residual''), positive for lines extending upwards from the curve (methylated CpG) and negative for lines extending downwards (unmethylated CpG), and take the average over the residual for all the CpGs in the interval that are covered by reads from this cell.
In this average, we perform shrinkage towards zero via a pseudocount (to trade bias for variance, see Methods for detail) in order to dampen the signal in cells with low coverage of the interval.

The average thus obtained, the shrunken mean of the residuals, is what we use to quantify this cell's (relative) methylation in the interval.
For a genome tiled into such intervals, we thus obtain a matrix, one row per cell, one column per interval, that can be used for downstream analysis, e.g., as input for PCA.
The signal-to-noise ratio in this matrix will be better than in a matrix obtained by simply averaging absolute methylation (0 or 1) over all the cells' covered CpG sites in a region.
The reason for this is that we reduce the variation in situations as the one depicted in Figure \ref{fig:smoothres}A, where the methylation of the reads might differ strongly even though there is no actual evidence for a difference between the two cells.

How should one obtain the ensemble average (the curved line in Fig.~\ref{fig:smoothres}B)?
A simple approach to get a value for a specific CpG would be to take all cells with read coverage for the CpG and use the fraction of these that show the CpG as methylated.
However, especially when only few cells offer coverage, these averages will be very noisy.
Therefore, we propose to smoothen using a kernel smoother, i.e., by performing a kernel-weighted average over the CpG site's neighborhood.
The kernel bandwidth (i.e., the size of the neighborhood to average over) is a tuning parameter; for the examples presented here we used 1000~bp.


% remove or shorten this section cause it's not really novel, methods like this are already implemented in e.g. the "missMDA" R package
%\subsection{PCA with imputation}
%The standard PCA algorithm cannot deal with missing data.
%Therefore, one would need to make intervals wide enough to ensure that every interval is covered by at least one read for most tiles and discard any interval that is not covered by any single cell.
%Hence, a naive analysis has to resort to a coarse tiling of the genome.
%To counter this, we propose a simple and straightforward way to deal with missing data in the input matrix to a PCA: In a first iteration, we replace each missing value in the centered input matrix with zero, then run the PCA.
%Then, only these zeroes are replaced by the value predicted by the PCA and the PCA is rerun.
%Convergence is typically achieved after two to five iterations.
%For details, see Methods.
%The possibility to perform PCA with missing data allows us to use shorter, and thus more intervals and hence provide the PCA with richer input data.

\subsection{Finding variably methylated regions}

\begin{figure}
    \includegraphics[width=\columnwidth]{figures/Fig_sliding.png}
    \caption{\small \textbf{Finding variably methylated regions (VMRs):} The chromosomes are divided up into overlapping windows (first five shown at the bottom), and for each window, the cells' methylation values are calculated as described and as depicted in Fig.~\ref{fig:smoothres}B.
Then, the variance of these values is calculated (each point represents one of the overlapping windows), a threshold (dashed line) is chosen such that a chosen quantile of windows have a variance exceeding the threshold.
Windows with above-threshold variance are merged if they overlap, yielding the ``variably methylated regions''.}
    \label{fig:vmr}
\end{figure}


Typically, some regions in a chromosome will have very similar methylation status in all cells, while other regions show variability in methylation across cells.
For instance, it is long known that CpG-rich promoters of housekeeping genes are unmethylated, and that a large proportion of the remaining genome is highly methylated regardless of cell type \citep{bird1986cpg}.
In contrast, DNA methylation at certain genomic features such as enhancers is more dynamic \citep{argelaguet2019gastru}, and thus more variable across cells.
Only the latter regions are of value for our goal of quantitating dissimilarity between cells.
We call these the variably methylated regions (VMRs). % maybe mention the parallel to highly variable genes in scRNAseq?

In the standard approach, one divides up (tiles) each chromosome into non-overlapping, equal-sized intervals, and quantitates the methylation of each tile.
Such rigid placing of interval boundaries is unlikely to be optimal: for example, a VMR might be much smaller than a tile and its signal will hence be drowned out by the larger part of uninformative CpG sites that are equal in all cells, when averaging over all the CpG sites in the tile.

Therefore, we propose the following approach (Fig.~\ref{fig:vmr}): Divide up the chromosome into many \emph{overlapping} windows, that start at regular multiples of a fixed, small, step size.
Quantify the methylation of each cell in each window by averaging the cell's methylation residuals over all CpGs in the window, as described above and depicted in Fig.~\ref{fig:smoothres}B.
Then, calculate for each window the variance of these values over all cells.
Select, say, the top 2\% windows with the highest variances and mark them as VMRs.
Wherever thus marked windows overlap or are separated by only a small gap, merge them into one larger VMR.
Then, calculate for each of these larger VMRs the methylation signal, as before, by averaging for each cell over the residuals of all contained CpG sites.

In this manner, we obtain a methylation matrix, with one row per cell and one column per VMR, that is, in a sense, richer in information and has better signal-to-noise ratio than the matrix obtained by the simple analysis sketched at the very beginning.
As we demonstrate below, a PCA performed on such a matrix provided a distance metric for the cells that contains more information on biological detail than one from a simpler analysis.

\subsection{Finding differentially methylated regions}

\begin{figure}
    \includegraphics[width=\columnwidth]{figures/Fig_DMRdetection.png}
    \caption{\small \textbf{Identification of differentially methylated regions (DMRs):}
    The chromosome is divided into windows overlapping by a small and fixed size.
    For each window, the cell’s methylation values are calculated as described and as depicted in Fig.~ \ref{fig:smoothres}B.
    Then the t-statistic of these values is obtained as a measure of methylation difference (each point represents one of the windows).
    Upper and lower thresholds (dashed lines) are chosen such that a defined quantile of windows have a t-statistic above or below the threshold.
    Windows exceeding either threshold are merged if they overlap, yielding the ``differentially methylated regions``.}
    \label{fig:dmrscan}
\end{figure}

A common task in the analysis of bulk bisulfite-sequencing data is the detection of differentially methylated regions (DMRs) between conditions, tissues, or cell types \citep{Hebestreit2013, dmrseq}.
As DNA methylation affects gene expression, DMRs can provide insights into the unique epigenetic and gene regulatory characteristics of cell types.
However, to date no approach to detect DMRs in scBS data has been reported.
To enable DMR detection in scBS data, we thus developed an approach that detects DMRs of variable size between two groups of cells, and controls the false discovery rate (Fig.~\ref{fig:dmrscan}):

Similar to the previously described VMR identification approach (Fig.~\ref{fig:vmr}), divide each chromosome into overlapping windows shifted by a small and fixed size (step size) and quantify the methylation of each cell in each window.
Next, instead of the variance, obtain the t-statistic % Welch t-test
as a measure of differential methylation between the two cell groups.
Identify the windows with the most extreme t-statistics, e.g. windows in the 2\% upper and lower tails.
Merge any overlapping windows in the upper tail into larger DMRs.
Repeat for windows in the lower tail, then re-calculate the t-statistic for each larger DMR.

To assess statistical significance of DMRs, repeat the same procedure on permutations of the scBS data, i.e. the same data set with randomly shuffled cell labels.
The DMR t-statistics obtained from permuted data are then used to estimate the false-discovery rate (FDR), yielding an adjusted p-value for each DMR.

\subsection{Utility of DMRs and VMRs} % not sure if we need a subsection here

While the primary purpose of VMRs is to provide better input for the distance calculations, their small size also means that they can be readily compared with genomic annotation, which aids in their functional interpretation.
Large 100~kb tiles will often contain multiple genes with distinct expression patterns, small VMRs located in known regulatory elements or gene bodies are more easily associated with individual genes and their associated function.
Thus, VMRs provide a useful starting point for examining the data with methods from functional genomics.
To enable a targeted characterization of epigenomic and gene regulatory differences between conditions or cell types, the same principles may also be applied to DMRs.


\section{Application and benchmarks}

To demonstrate the value of our proposed improvements, we obtained two previously published scBS data sets and re-analyzed them using our methods.

\phantomsection
\subsection{Correlating VMR methylation with gene expression}

The first data set used to test our methods is a single-cell multi-omics data set generated by \citet{kremer_scnmt}.
This data set comprises the methylomes of 1566 cells isolated from mouse forebrains, as well as matched single-cell transcriptomes of the same cells.
Among these cells are distinct cell types such as oligodendrocytes, oligodendrocyte precursor cells (OPCs) and endothelial cells, as well as cellular sub-states that are part of the continuous neural stem cell differentiation trajectory.
To assess whether our VMR detection method captures genomic intervals that are biologically meaningful, we probed whether their methylation level correlates with the expression of nearby genes.
Compared to promoter methylation, the obtained correlation coefficients are more extreme, indicating that VMR methylation often is a better predictor of gene expression than promoter methylation (Fig.~\ref{fig:correlation}A).
Indeed, a gene-wise comparison revealed many genes whose expression is correlated with the nearest VMR, but not with promoter methylation (Fig.~\ref{fig:correlation}B).
One such example gene, Htra1, is depicted in Fig.~\ref{fig:correlation}C.
While the promoter of this gene is lowly methylated regardless of gene expression, a VMR located downstream of the promoter is lowly methylated in cells with high Htra1 expression.

\begin{figure}
    \begin{center}
    \includegraphics[width=.95\columnwidth]{figures/Fig_correlation.png}
    \end{center}
    \caption{\small \textbf{Correlation of DNA methylation and gene expression}.
    \textbf{(A)} Distribution of Pearson correlations between gene expression and promoter methylation (black) and gene expression and methylation of the nearest VMR (red).
    \textbf{(B)} Correlation of each gene's RNA expression values to methylation (shrunken mean of residuals) at promoters (y-axis), or to methylation of the nearest VMR (x-axis).
    Promoters are defined as intervals \textpm2~kb around the TSS.
    \textbf{(C)} Mean methylation near the gene Htra1 (ENSMUSG00000006205).
    Cells are assigned to 4 groups based on Htra1 expression (group 0: cells with no Htra1 expression, groups 1-3: cells that express Htra1, divided into three equally large groups with group 3 having the highest expression).
    Data from \citet{kremer_scnmt}.}
    \label{fig:correlation}
\end{figure}


\subsection{Effects of our methods on cell type identification}

\begin{figure}
    \begin{center}
        \includegraphics[width=\columnwidth]{figures/Fig_benchmark.png}
    \end{center}
    \caption{\small \textbf{Benchmark of our methods on single-cell multi-omic data of cells of the murine forebrain}.
    \textbf{(A)} Mean neighbor score obtained when analyzing data from \citet{kremer_scnmt} with different combinations of methods.
    Either 100~kb genomic tiles or VMRs were subjected to PCA and UMAP.
    Tiles or VMRs were pre-processed as proposed in this work (shrunken mean of residuals subjected to iterative PCA, labeled "pre-processing: ours") or using conventional methods (PCA on lightly imputed methylation fractions, based on \citep{luo2017single}).
    The full 1566-cell data set was sub-sampled to simulate smaller data sets (x axis).
    A higher neighbor score implies better separation of cell types inferred from single-cell transcriptomes of the same cells \textbf{(B)}.
    \textbf{(C, D)} Exemplary UMAPs obtained when analyzing the data set with our proposed methods (C) or conventional methods based on genome tiling (D).}
    \label{fig:score}
\end{figure}

To assess the impact of our proposed methods, we next tested whether our methods improve the ability to distinguish cell types and cell states (Fig.~\ref{fig:score}).
To this end, we obtained cluster labels based on the single-cell transcriptomes from \citet{kremer_scnmt} (Fig.~\ref{fig:score}B).
We consider these cell labels as ground truth and next tested whether we are able to distinguish the same groups of cells based on their methylomes.
To do this, we subjected the methylomes to the standard scBS workflow:
defining a set of genomic intervals, quantifying methylation in these intervals, and subjecting the resulting matrix to PCA.
To determine genomic features for distance calculation, we either tiled the genome into 100~kb bins as is commonly done, or we used our VMR detection approach.
To quantify methylation in these intervals, we either averaged in these intervals, obtaining methylation percentages ("classic" pre-processing), or we calculated the shrunken mean of residuals.
Visual inspection of the resulting UMAPs revealed that our proposed combination of methods results in more clearly separated cell types, compared to a UMAP obtained with default analysis methods (Fig.~\ref{fig:score}C,D).
While all cells form a continuous point cloud when using default methods, our improvements led to a clear separation of oligodendrocytes, OPCs and endothelial cells.
Furthermore, even cellular sub-types of cells in the continuous neural stem cell lineage were partially separated.

To quantify this performance gain in a more rigorious manner, we used a score that quantifies whether cells were placed in a neighborhood consisting of cells with the same cell type in 15-dimensional PC space ("neighbor score", see \ref{methods:score}).
Since scBS protocols are costly and labor-intensive, only few laboratories are currently able to obtain thousands of single cell methylomes.
To simulate smaller data sets, we thus sub-sampled the 1566-cell data set into smaller data sets, analyzed them with different method combinations, and calculated the mean neighbor score for each (Figure \ref{fig:score}A).
The results demonstrate that the forebrain cell clusters are more cleanly separated when using our proposed methods.
This performance gain was observed also in smaller data sets, although cell cluster separation generally becomes more difficult in those.
Both the use of VMRs instead of genomic tiles, as well as the use of the mean shrunken residuals over methylation averages improved cell cluster separation.


\subsection{Identification of biologically relevant DMRs}

\begin{figure*}[t]
    \begin{center}
        \includegraphics[width=.8\textwidth]{figures/Fig_DMRs.png}
    \end{center}
    \caption{\small \textbf{identified neuroblast and oligodendrocyte DMRs have a putative regulatory role:} \textbf{(A)} Difference in methylation between oligodendrocytes and neuroblasts of DMRs plotted against the number of cells contained in a DMR.
Colored by p-value.
\textbf{(B)} Mean methylation of cells in the DMR covering MBP.
Line indicates median.
\textbf{(C)} Mean methylation of reads at individual sites along the DMR covering MBP.
\textbf{(D)} GREAT \citet{mclean2010great} was used for GO-term analysis.
The left side shows GO-terms for genes near the 3000 most significant neuroblast DMRs sorted by GO-enrichment p-value and t-statistic (green arrow) and the right side for oligodendrocyte DMRs (yellow arrow) respectively.
    }
    \label{fig:dmr}
\end{figure*}

As a demonstration and to assess the ability of the presented approach to detect relevant DMRs, the method was used to find regions differing in CpG methylation between vSVZ neuroblasts and oligodendrocytes.
Fig.~\ref{fig:dmr} gives an overview of all identified DMRs and the relation of the adjusted p-value to the number of cells included in a DMR, and the difference in methylation between the cell types.
Logically, if a DMR contains few cells, the methylation difference between cell types must be higher to obtain a low (and thus more significant) p-value.
A low p-value is maintained, although the methylation difference decreases if the number of cells in the DMR increases.


An oligodendrocyte DMR stretching across Mbp was chosen to assess the methylation patterns of an individual DMR (Fig.~\ref{fig:dmr}B-C).
Mbp is highly expressed in oligodendrocytes as it is essential for the myelination of neuronal axons.
The DMR is highly methylated in neuroblasts and lowly methylated in oligodendrocytes (Fig.~\ref{fig:dmr}B-C), including more substantial demethylation in oligodendrocytes at the position of Mbp (Fig.~\ref{fig:dmr}C).
Low methylation at this locus supports the expression of Mbp.


Nevertheless, this is not the only indicator of the biological relevance of discovered DMRs.
GO-term enrichment shows that neuroblast DMRs occur near genes involved in the regulation of neurogenesis, neuron differentiation, and the positive regulation of neural precursor cell proliferation, all of which are essential for neuroblasts (Fig.~\ref{fig:dmr}D).
On the other side, oligodendrocyte-defining DMRs are located near genes with implications for myelin sheath, myelination, and axon ensheathment and thus are linked to oligodendrocytes.
Therefore, DMRs occur in the proximity of genes associated with oligodendrocyte or neuroblast function, respectively and are implied to regulate gene expression and contribute to fundamental cell type-specific function, which is supported by the repressive effect of CG methylation on gene expression.


\section{The \texttt{scbs} Python package}

\begin{figure*}[t]
    \begin{center}
        \includegraphics[width=.8\textwidth]{figures/Fig_workflow.png}
    \end{center}
    \caption{\small \textbf{Overview of the functionalities implemented in the \texttt{scbs} package.}
        \texttt{scbs prepare} parses methylation files produced by common bisulfite sequencing mappers and stores their contents in a compressed format optimised for efficient access to genomic intervals.
        \texttt{scbs} also produces cell-wise summary statistics and quality plots (here: average methylation around the transcription start site) that are used to detect low-quality cells.
        These cells can be discarded with \texttt{scbs filter}.
        To obtain a methylation matrix, similar to the count matrices used in scRNA-seq, the user must first decide in which genomic intervals methylation should be quantified.
        The user may either provide genomic regions of \emph{a priori} interest, or they may choose to discover VMRs (variably methylated regions) in the data with \texttt{scbs scan}.
        The resulting methylation matrix can then be used for downstream analysis such as cell clustering and dimensionality reduction.
        Identified groups of cells (e.~g. clusters) can then be scanned for DMRs with \texttt{scbs diff}.
    }
    \label{fig:workflow}
\end{figure*}

We have implemented the approach just described in a Python package with a command line interface, called \texttt{scbs}, which also offers a number of other functionalities for the analysis of scBS data (Figure \ref{fig:workflow}).
The starting point of such an analysis are methylation files generated by tools such as Bismark \citep{bismark} or methylpy \citep{methylpy}.
For each cell, these tools produce a text file that lists the methylation status of all CpG sites.
Since it is inconvenient to work with hundreds or thousands of text files, \texttt{scbs} provides the \texttt{prepare} command which parses these methylation files and stores their content in a compressed format that enables efficient access to all CpG sites of the genome.
\texttt{scbs prepare} is flexible and accepts all tabular input formats including Bismark .cov-files, methylpy .allc-files, or custom user-defined formats.

In brief, we store methylation of each chromosome in a matrix where each column represents a cell and each row represents a base pair.
Methylated and unmethylated sites are coded as $+1$ and $-1$, respectively.
Since the vast majority of values in this matrix are missing due to the sparsity of scBS data (and because rows for base-pairs not corresponding to a CpG site contain no data), we encode missing values as zero and then store the data in Compressed Sparse Row (CSR) format.
This format does not explicitly store zeroes (here: missing values) and is optimized for row-wise access, which results in significant compression and allows fast access to the methylation status of genomic intervals.

\texttt{scbs prepare} also computes a number of summary statistics for each cell, including the mean genome-wide methylation level and the number of observed CpG sites, i.e.
the number of CpG sites that have sequencing coverage.
These summary statistics can be used to detect cells with poor quality.
The quality of each single cell methylome can furthermore be inspected with \texttt{scbs profile}, which computes the average methylation profile of a set of user-defined genomic regions such as transcription start sites (TSS) at single-cell resolution.
The TSS profile is a useful quality control plot since methylation shows a characteristic dip roughly \textpm1~kb around the transcription start site in mammalian genomes.
Cells that do not show this pattern, or cells with few observed CpG sites, may then be discarded with \texttt{scbs filter}.


After quality control, the user has access to the genome-wide VMR detection approach described earlier, using the \texttt{scbs scan} command.
This produces a BED file that lists the genome coordinates of VMRs.
Similarly, the user can identify DMRs after quality control with \texttt{scbs diff} using a text file listing the cells’ group assignment as input.
The command produces a BED file listing the coordinates of the determined DMRs together with the t- statistic, the corresponding cell group, and the adjusted p-value.
To finally obtain a methylation matrix analogous to a scRNA-seq count matrix, this BED file can be used as input for \texttt{scbs matrix}, which quantifies methylation at genomic intervals in all cells.
The command produces both the percentage of methylated CpG sites, as well as our proposed methylation measure (shrunken mean of the residuals) that is more robust to variation in sequencing coverage or stochastic differences in read position between cells.
We note that both \texttt{matrix} and \texttt{profile} accept any valid BED file as input, which means that the user can quantify and visualize methylation at any genomic feature of interest, such as promoters, enhancers or transcription factor binding sites, in single cells.
The obtained methylation matrix can then serve as input for common single-cell analysis methods such as dimensionality reduction and cell clustering.

\subsection{Availability}

\texttt{scbs} is free and open source.
The package is available on the Python Package Index (PyPI) and can be installed with the Python package installer pip.
The source code is hosted at \href{https://github.com/LKremer/scbs}{https://github.com/LKremer/scbs}, where we also provide detailed documentation including a tutorial that demonstrates a complete \texttt{scbs} analysis on an example data set.




\section{Discussion and Conclusion}

Here, we have proposed an improved strategy to preprocess single-cell bisulfite sequencing data.
Based on the observation that incomplete read coverage of a genomic interval can lead to inaccurate methylation estimates, we suggest a scoring scheme that is aware of a reads local context rather than just treating all reads in an interval alike.

Furthermore, we show a way to pinpoint minimal regions of high variability across cells, which we call variably methylated regions (VMRs).
By re-analyzing two published data sets, we could show that these improvements to data preprocessing help to increase signal and decrease noise, resulting in a more informative intermediate-dimensional representation of the data.
As examples of practical benefits, we demonstrate that our preprocessing allows for better distinction of cell sub-types, especially for smaller data sets with a limited number of cells.
In addition to this improvement of global signal-to-noise ratio (with regards to cell dissimilarity quantification), we also show that the determination of VMRs improve biological interpretability by pinpointing the regions that actually change and hence have a putative regulatory role.

We also presented and published an open source software tool, called \texttt{scbs}, that provides an easy-to-use implementation of the described algorithms.
The package also offers functionality for all other steps of data preprocessing that are needed to get from the output of methylation callers such as Bismark and methylpy to a cell\texttimes region matrix that is suitable as input for standard analysis tools that work in a reduced-dimensionality setting, thus allowing the user to use, e.g., well established tools from scRNA-seq analysis, including visualization tools such as t-SNE and UMAP, Leiden/Louvain clustering, pseudotime trajectory analyses, etc.
By offering these functions, \texttt{scbs} bridges a gap in the chain of existing tools that so far hindered practitioners in their data analysis.

In addition, by pinpointing exact regions of interest, the VMR functionality provides useful input for genomics-style analyses.
Depending on the research question at hand, individual VMRs can be related to nearby genomic features such as gene bodies or known regulatory elements, tested for methylation differences between groups of cells, or subjected to gene ontology enrichment or motif enrichment.

In conclusion, we have presented powerful improvements to scBS data preprocessing and a software tool that implements these and bridges a gap in existing tool chains.


\vspace{1.4ex}
\noindent\hfil\rule{.6\columnwidth}{.2pt}\hfil

\phantomsection
\subsection{Acknowledgments}

S.\ A.\ and L.\ P.\ M.\ K.\ acknowledge funding by the Klaus Tschira Foundation.
L.\ P.\ M.\ K., L.\ K., S.\ C.\ and A.\ M.-V.\ acknowledge funding from the European Commission via ERC grant 771376.
We thank Alexey Uvarovskii for code refactoring and helpful suggestions on the source code.

\phantomsection
\section{Methods}

\phantomsection
\subsection{Raw data}

We write $x_{ij}$ for the methylation status of CpG $i$ in cell $j$.
The index $i$ runs over all CpG positions present in the genome, the index $j$ over all cells in the assay.
We write $x_{ij}=0$ if position $i$ was found to be unmethylated in cell $j$ by bisulfite sequencing, $x_{ij}=1$ if it was methylated, and $x_{ij}=\text{NA}$ if position $i$ was not covered by reads from cell $j$ and the methylation status is therefore not available (``NA''),

These values can be readily obtained from single-cell bisulfite sequencing data using tools like Bismark.

If multiple reads from the same cell cover a position, these will typically be PCR duplicates of each other and hence agree.
Of course, the two alleles of a CpG may rarely differ in their methylation status.
While it is in principle possible that one obtains discordant reads stemming from the same position on both the paternal and the maternal chromosomes of the same cell, this is so unlikely that we can ignore the case.
Hence, whenever the methylation caller reports multiple reads covering the same position in the same cell, we set $x_{ij}$ to 0 or 1 whenever all reads agree.
When there is disagreement, we put $x_{ij}=\text{NA}$ by default, or optionally follow the majority of reads whenever possible.

For later use: We write $C$ for the set of all cells in the assay (i.e., $C$ is the index set for the cell indices $j$).
Moreover,
we define $C_i\subset C$ as the set of all those cells $j$ that have reads covering position $i$,
$$ C_i=\{j\in C: x_{ij}\neq\text{NA}\}.$$
Conversely, we define $G_j$ as the set of all the CpG positions $i$ covered by reads from cell $j$ 
$$ G_j=\{i: x_{ij}\neq\text{NA}\}.$$

\subsection{Data storage}

The function \texttt{scbs prepare} reads a set of methylation files (e.g.
produced by Bismark) and produces a file that stores the matrix $x$ in a space-efficient format, as follows: $x$ is represented as a SciPy sparse matrix \citep{SciPy}, encoding the actual values 0, 1, and NA as -1, 1, and 0, respectively.
Coding NA (the most common value) as 0 leverages SciPy's sparse matrix storage.
In all the follows here, any mention of $x$ will, however, always mean the encoding as $x_{ij}\in\{0,1,\text{NA}\}$.

\subsection{Smoothing}

For each CpG position $i$, we write 
$$\overline{x}_i=\langle x_{ij} \rangle_{j\in C_i} = \frac{1}{|C_i|}\sum_{j\in C_i} x_{ij}$$ 
for the average methylation at position $i$, where $\langle\cdot\rangle$ denotes averaging, and the average runs over all the cells $j\in C_i$, i.e.
over those cells for which methylation data is available for position $i$.

%\todo{NOTE: Here, we could already do shrinkage and use $\frac{1}{|C_i|+1}\sum_i x_{ij}$ instead. But we don't do that, right?}

We then run a kernel smoother over these per-position averages to obtain the smoothed averages $\tilde x_i$.
Specifically, we use
\[ \tilde x_i = \frac{\sum_{i'} \overline x_{i'}\, k_h(d_{ii'})}{\sum_{i'} k_h(d_{ii'})},\]
i.e., $\tilde x_i$ is the weighted average over the per-position averages $\overline{x}_{i'}$, taken over the CpG sites $i'$ in the neighborhood of $i$, and weighted using a smoothing kernel $k_h$ with bandwidth $h$.
Here, $d_{ii'}$ is the distance between CpG positions $i$ and $i'$, measured in base pairs, $h$ is the smoothing bandwidth in base pairs (by default, $h=1000$), and $k_h$ is the tricube kernel,

\[ k_h(d) = \left\{
\begin{aligned}
    &\left(1-|d/h|^3\right)^3 &\text{for } |d|<h \\
    &\,0 &\text{otherwise}.
\end{aligned}
\right.
\]

\subsection{Methylation for an interval}

Next, we discuss averaging methylation over a range of CpG sites.

Given an interval $I$ on the chromosome, we wish to quantify the average methylation $m_{Ij}$ of the CpG sites within the interval for cell $j$.
If we interpret $I$ as the set of CpG positions $i$ in the interval, we may write

$$ m_{Ij} = \left< x_{ij} \right>_{i\in I\cap G_j}.$$

Here, the average runs over all those sites $i$ that lie within the interval $I$ and are covered by reads from cell $j$.

If we wish to compare cells, it can be helpful to center this quantity by subtracting its average:

$$ z_{Ij} = m_{Ij} - \langle m_{Ij'}\rangle_{j'\in C}.$$

As an alternative, we suggest to consider the residuals of the individual CpG methylation values $x_{ij}$ from the smoothed average $\tilde x_i$,
$$ r_{ij} = x_{ij} - \tilde x_i, $$
and averaging over these, obtaining
\begin{equation} 
r_{Ij} = \frac{\scriptsize{1}}{\scriptsize{|I\cap C_i|+1}}\sum_{i\in I\cap C_i}\left( x_{ij} - \tilde x_i \right).
\label{avgres}
\end{equation}

This is a shrunken average, with denominator $n+1$.
This extra pseudocount has the effect of shrinking the value towards the ``neutral'' value 0, with the shrinkage becoming stronger if the data is ``weak'' because the number $|I\cap C_i|$ of positions covered by reads from cell $j$  is low.
In the extreme case of none of the reads from cell $j$ covering $I$, the sum becomes 0 and the denominator 1, i.e.
$r_{Ij}=0$ in this case.

\subsection{Finding variably methylated regions}

For any interval $I$, we denote by $v_I$ the variance of its residual averages $r_{Ij}$:

$$ v_I = \frac{\scriptsize{1}}{\scriptsize{|C_I|}}\left( r_{Ij} - \langle r_{Ij'}\rangle_{j'\in C_I} \right)^2,$$

where the average runs only over the set $C_I=\bigcup_{i\in I}C_i$ of those cells $j$ which have reads covering interval I.

To find VMRs, we define intervals $I_1, I_2, ...$, all of the same width, and with step-wise increasing starts, then calculate $v_1, v_2, ...$ for these intervals.
We then mark the intervals with the 2\% highest variances.
We take the union of all these intervals, split the union into connected components, and call each component a VMR.
Putting that last step in other words: We take all the intervals with variance in the top 2-percentile, fuse intervals that overlap and call the regions thus obtained the VMRs.

\subsection{Calculating cell to cell distances}

Given a set, $\mathcal{V}=\{I^\text{v}_1,I^\text{v}_2,\dots\}$, of intervals corresponding to VMRs, we get a relative methylation fraction $r_{ij}$ for each VMR $I^\text{v}_i$ and each cell $j$ from Eq.\ (\ref{avgres}).
The matrix thus obtained can then be centered and used as input for a PCA.
If we calculate the top $R$ principal components, we thus obtain for each cell $j$ an $R$-dimensional principal component vector $\mathbf{x}^\text{P}_j$.
For any two cells $j$, $j'$, we use the Euclidean distance $\|\mathbf{x}^\text{P}_j - \mathbf{x}^\text{P}_{j'}\|$ as the measure of dissimilarity of the two cells.
Thus, the matrix of PC scores can be used as input to dimension reduction methods like t-SNE or UMAP, and to clustering methods like the Louvain or Leiden algorithm, which require such a matrix as input to the approximate nearest neighbor (ANN) finding algorithm that is their first step.

\subsection{PCA with imputation}

Whenever a region is not covered by any read in a cell, the corresponding data entry in the input data matrix for PCA will be missing.
The standard approach to calculate PCA (commonly done using the IRLBA algorithm, \citep{Baglama2005}) is not suited to deal with missing data.
We circumvent this issue by simply using the PCA itself to impute the missing value.

Let us write $A$ for the matrix to which the PCA is to be applied, with the features (here: regions) represented by the matrix rows.
The matrix has already been centered, i.e., $\sum_i a_{ij}=0$.
To establish notation, we remind the reader that performing a PCA on A means finding the singular value decomposition (SVD), $A=UDR^\top$, with $D$ diagonal and $U$ and $R$ orthonormal.
The PCA scores are contained in $X=UD$, the loadings in $R$.
To reconstruct the input data $A$ from the PCA representation, one may use $A=XR^\top$, i.e., $a_{ij}=\sum_r x_{ir} r_{jr}$, where the equation is exact if $r$ runs over all principal components and approximate if it is truncated to the leading ones.

Our iterative imputation strategy is now simply the following: we first replace all missing values in the row-centered input matrix $A$ with zeroes and perform the (truncated) PCA.
Then, we use the PCA predictions for the missing values, i.e., the $a_{ij}=\sum_r x_{ir} r_{jr}$, as refined stand-ins for the missing values in $A$ and run PCA once more.
This can now be iterated until convergence, which is typically achieved after at most 4--5 rounds of PCA.

\subsection{Re-analysis of published scBS datasets}
To re-analyze scBS data sets by \citet{luo2017single} and \citet{argelaguet2019gastru}, the genome was first annotated with genomic intervals that were either obtained by tiling the genome into non-overlapping 100~kb tiles, or by acquiring VMRs with \texttt{scbs scan} using the current default options (bandwidth $=2000$, stepsize $=10$, variance threshold $= 2\%$).
For all regions (VMRs and 100~kb tiles), the mean methylation fractions and the shrunken means of residuals were obtained with \texttt{scbs matrix}.
Regions containing less than four measured CpGs, or that were measured in less than 10\% of cells, were discarded.
To select $n$ intervals for further analysis, the top $n$ intervals were picked based on the number of cells that the region was measured in.

For dimensionality reduction, missing values were imputed with the iterative PCA described in the previous section.
To visualize data, UMAP was performed on the final 15 principal components.

\subsection{Nearest neighbor count score}
To assess performance of our methods, we processed mouse CpG methylation data from \citet{luo2017single} with several method combinations and scored each result.
The score evaluates the ability of each method combination to separate cell types in 15-dimensional PCA space, based on their CpG methylation.
Each score varies between 0 and 1, where higher scores reflect better separation.

Our nearest-neighbor-count score $\text{nnc}_k$, is based on the $\Gamma$-score \citep{Kireeva_2014} and is calculated as follows.
For each cell $j$, we count how many of its $k$ nearest neighbors have been assigned to the same cluster as cell $j$.
We denote this count $a^k_j$.
We consider a cell as ``well-assigned'' if $a^k_j>\theta k$, where we have chosen $\theta=0.9$.
In other words, a cell is ``well-assigned'' if at least 90\% of its $k$ nearest neighbors have been assigned to the same class as the cell under consideration.

The score $\text{nnc}_k$ is then simply the fraction of cells in the data set that are well-assigned in this sense.

\vspace{1.4ex}
\noindent\hfil\rule{.6\columnwidth}{.2pt}\hfil


{\small \bibliography{scbs}}

\end{document}